File: chess_cpp/validators/KingUnderAttackValidator.cpp
//
// Created by Hrach Khachatryan on 12/15/19.
//

#include "KingUnderAttackValidator.h"
#include "../piecies/King.h"
#include "../utils/MessageHandler.h"

bool KingUnderAttackValidator::validate(Board *board, Spot *start, Spot *end) {
    bool isWhite = start->getPiece().get()->isWhite();
    std::shared_ptr<Piece> endPiece = end->getPiece();
    end->setPiece(start->getPiece());
    start->setPiece(nullptr);

    Spot* kingSpot = board->getKingSpot(isWhite);
    King king = (King)kingSpot->getPiece().get();
    bool kingIsUnderAttack = king.isKingUnderAttack()->isUnderAttack();
    start->setPiece(end->getPiece());
    end->setPiece(endPiece);

    return !kingIsUnderAttack;
}

std::string KingUnderAttackValidator::getNotValidMessage() {
    return MessageHandler::getNoMessage();
}

File: oop_chess/src/chess/validators/KingUnderAttackValidator.java
package chess.validators;

import chess.Board;
import chess.Spot;
import chess.piecies.King;
import chess.piecies.Piece;
import chess.utils.MessageHandler;

public class KingUnderAttackValidator extends PieceMoveValidator {

    @Override
    public boolean validate(Board board, Spot start, Spot end) {
        boolean isWhite = start.getPiece().isWhite();
        Piece endPiece = end.getPiece();
        end.setPiece(start.getPiece());
        start.setPiece(null);

        Spot kingSpot = board.getKingSpot(isWhite);
        King king = (King)kingSpot.getPiece();
        boolean kingIsUnderAttack = king.isKingUnderAttack().isUnderAttack();
        start.setPiece(end.getPiece());
        end.setPiece(endPiece);

        return !kingIsUnderAttack;
    }

    @Override
    public String getNotValidMessage() {
        return MessageHandler.getNoMessage();
    }
}
